File: tests/drivers/go/v5/write_routing.go
package main

import (
	"fmt"
	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
)

func create_message(uri string) {
	username := ""
	password := ""

	// Connect to Memgraph
	driver, err := neo4j.NewDriver(uri, neo4j.BasicAuth(username, password, ""))
	if err != nil {
		panic(err)
	}
	defer driver.Close()

	session := driver.NewSession(neo4j.SessionConfig{AccessMode: neo4j.AccessModeWrite})
	defer session.Close()

	greeting, err := session.WriteTransaction(func(transaction neo4j.Transaction) (interface{}, error) {
		result, err := transaction.Run("CREATE (n:Greeting) SET n.message = $message RETURN n.message", map[string]interface{}{
			"message": "Hello, World!",
		})
		if err != nil {
			return nil, err
		}

		if result.Next() {
			return result.Record().Values[0], nil
		}

		return nil, result.Err()
	})

	if err != nil {
		panic(err)
	}

	fmt.Println(greeting)
}

// Test checks that you can use bolt+routing for connecting to main and coordinators for writing.
func main() {
	fmt.Println("Started running main_route.go test")
	create_message("neo4j://localhost:7687") // instance_1
	create_message("neo4j://localhost:7690") // coordinator_1
	create_message("neo4j://localhost:7691") // coordinator_2
	create_message("neo4j://localhost:7692") // coordinator_3
	fmt.Println("Successfully finished running main_route.go test")
}

File: tests/drivers/java/v5_8/src/main/java/memgraph/WriteRouting.java
package memgraph;

import static org.neo4j.driver.Values.parameters;

import java.util.*;
import java.util.concurrent.TimeUnit;
import org.neo4j.driver.AuthTokens;
import org.neo4j.driver.Config;
import org.neo4j.driver.Driver;
import org.neo4j.driver.GraphDatabase;
import org.neo4j.driver.Result;
import org.neo4j.driver.Session;
import org.neo4j.driver.Transaction;
import org.neo4j.driver.TransactionWork;
import org.neo4j.driver.exceptions.ClientException;
import org.neo4j.driver.exceptions.TransientException;

public class WriteRouting {
    private Driver driver;

    private void createMessage(String uri) {
        driver = GraphDatabase.driver(uri, AuthTokens.basic("", ""));
        try (Session session = driver.session()) {
            String greeting = session.writeTransaction(tx -> {
                var result = tx.run("CREATE (n:Greeting) SET n.message = $message RETURN n.message",
                                    parameters("message", "Hello, World!"));
                if (result.hasNext()) {
                    return result.single().get(0).asString();
                }
                throw new RuntimeException("No result found.");
            });
            System.out.println(greeting);
        }
    }

    public static void main(String... args) {
        System.out.println("Started running WriteRoutingTest...");
        WriteRouting greeter = new WriteRouting();
        greeter.createMessage("neo4j://localhost:7690"); // coordinator_1
        greeter.createMessage("neo4j://localhost:7691"); // coordinator_2
        greeter.createMessage("neo4j://localhost:7692"); // coordinator_3
        System.out.println("All good!");
    }
}
